datadog_checks_base/datadog_checks/base/stubs/similar.py

```diff
@@ -47,7 +47,7 @@ def _build_similar_elements(expected_element, submitted_elements):
         raise NotImplementedError("Invalid type: {}".format(expected_element))
 
     similar_elements = []
-    for _, metric_stubs in iteritems(submitted_elements):
+    for _, metric_stubs in iteritems(list(submitted_elements)):
```

Comment:
Hmm, did you mean `list(iteritems(submitted_elements))`? Failures [like this one](https://github.com/DataDog/integrations-core/actions/runs/5204887302/jobs/9389714686?pr=14673#step:12:1593) happen because `six.iteritems` expects a dict not a list as argument.

Reply:
Tbh I'm not sure I understand the need for this change. We're not actually modifying `submitted_elements` in the for-loop. Would you mind reverting it and running the tests?

Reply:
👋 we were seeing failures like this one:
```
=================================== FAILURES ===================================
_______________________ test_pg_stat_statements_dealloc ________________________
tests/test_statements.py:1809: in test_pg_stat_statements_dealloc
    aggregator.assert_metric("postgresql.pg_stat_statements.dealloc", value=1, tags=expected_tags)
../datadog_checks_base/datadog_checks/base/stubs/aggregator.py:368: in assert_metric
    self._assert(condition, msg=msg, expected_stub=expected_metric, submitted_elements=self._metrics)
../datadog_checks_base/datadog_checks/base/stubs/aggregator.py:409: in _assert
    new_msg = "{}\n{}".format(msg, build_similar_elements_msg(expected_stub, submitted_elements))
../datadog_checks_base/datadog_checks/base/stubs/similar.py:19: in build_similar_elements_msg
    similar_metrics = _build_similar_elements(expected, submitted_elements)
../datadog_checks_base/datadog_checks/base/stubs/similar.py:50: in _build_similar_elements
    for _, metric_stubs in iteritems(submitted_elements):
E   RuntimeError: dictionary changed size during iteration
- generated xml file: /home/runner/work/integrations-core/integrations-core/postgres/.junit/test-unit-py3.8-14.0$TOX_ENV_NAME.xml -
=========================== short test summary info ============================
FAILED tests/test_statements.py::test_pg_stat_statements_dealloc - RuntimeError: dictionary changed size during iteration
============= 1 failed, 199 passed, 3 skipped in 123.64s (0:02:03) =============
```

which we could not re-produce locally. Seems like a flakey test? Looks like mid-loop there there is some call to `aggregator.<metric>`, which must be happening in a thread. That adds to the `Aggregator._metrics` list, which is a class attribute: [https://github.com/DataDog/integrations-core/blob/6f3ef03df4ac44f6219ee1db1da5e104[…]9f7/datadog_checks_base/datadog_checks/base/stubs/aggregator.py](https://github.com/DataDog/integrations-core/blob/6f3ef03df4ac44f6219ee1db1da5e1047a0d49f7/datadog_checks_base/datadog_checks/base/stubs/aggregator.py#L239)

Originally, she fixed this by making a copy of the elements first and then iterated over them, would that be acceptable?

Reply:
Your diagnosis sounds plausible, but I'm having trouble finding suspects right now, will take a look at this tmr with a fresher head.

cc @ofek as he's more familiar with the aggregator rig and the extent to which its global state is susceptible to these kinds of problems.

Reply:
My take on this is that the aggregator is not designed to be thread-safe, and from a quick look (I'm not familiar with how the dbm integrations use threading), the following lines inside the `check()` method spawn threads that are not joined within the method itself, causing a race condition for the aggregator: https://github.com/DataDog/integrations-core/blob/22899496fd0908344f1db0920d55554645b85578/postgres/datadog_checks/postgres/postgres.py#L737-L739

A couple of thoughts on that (assuming my assessment was correct):
- I don't think it's a good idea to patch the aggregator stub willy-nilly to make it more thread-safe. I think that if it being thread-safe becomes a requirement, some thought and design needs to be put into it. Otherwise we end up with something that's kinda thread-safe sometimes but nobody's sure when, and with undocumented tweaks to the code for the sake of thread safety.
- Spawning threads and not waiting for them to finish within the same test screams "flaky" to me, more so if you're making assertions on behavior that relies on those threads (since they run async, the execution order is completely nondeterministic.

Reply:
I agree with Alex's assessment, the code should wait for thread completion before test assertions occur